micro_http/src/common/headers.rs

```diff
@@ -13,63 +12,227 @@ pub enum Header {
+            Header::TransferEncoding => b"Transfer-Encoding",
+        }
+    }
+
+    fn try_from(string: &[u8]) -> Result<Self, RequestError> {
+        if let Ok(s) = String::from_utf8(string.to_vec()) {
+            match s.trim() {
+                "Content-Length" => Ok(Header::ContentLength),
+                "Content-Type" => Ok(Header::ContentType),
+                "Expect" => Ok(Header::Expect),
+                "Transfer-Encoding" => Ok(Header::TransferEncoding),
+                _ => Err(RequestError::InvalidHeader),
+            }
+        } else {
+            Err(RequestError::InvalidHeader)
         }
     }
 }
 
-/// Wrapper over the list of headers associated with a Request/Response.
+/// Wrapper over the list of headers associated with a Request that we need
+/// in order to parse the request correctly and be able to respond to it.
+///
+/// The only `Content-Type`s supported are `text/plain` and `application/json`, which are both
+/// in plain text actually and don't influence our parsing process.
+///
+/// All the other possible header fields are not necessary in order to serve this connection
+/// and, thus, are not of interest to us. However, we still look for header fields that might
+/// invalidate our request as we don't support the full set of HTTP/1.1 specification.
+/// Such header entries are "Transfer-Encoding: identity; q=0", which means a compression
+/// algorithm is applied to the body of the request, or "Expect: 103-checkpoint".
 #[derive(Debug)]
 pub struct Headers {
-    headers: HashMap<Header, String>,
+    /// The `Content-Length` header field tells us how many bytes we need to receive
+    /// from the source after the headers.
+    content_length: i32,
+    /// The `Expect` header field is set when the headers contain the entry "Expect: 100-continue".
+    /// This means that, per HTTP/1.1 specifications, we must send a response with the status code
+    /// 100 after we have received the headers in order to receive the body of the request. This
+    /// field should be known immediately after parsing the headers.
+    expect: bool,
+    /// `Chunked` is a possible value of the `Transfer-Encoding` header field and every HTTP/1.1
+    /// server must support it. It is useful only when receiving the body of the request and should
+    /// be known immediately after parsing the headers.
+    chunked: bool,
 }
 
 impl Headers {
     /// By default Requests are created with no headers.
     pub fn default() -> Headers {
         Headers {
-            headers: HashMap::new(),
+            content_length: 0,
+            expect: false,
+            chunked: false,
         }
     }
 
-    /// Adds a new header to the list.
-    pub fn add(&mut self, header: Header, value: String) {
-        self.headers.insert(header, value);
+    /// Expects one header line and parses it, updating the header structure or returning an
+    /// error if the header is invalid.
+    ///
+    /// # Errors
```

Comment:
Please also add an issue to fix the errors returned by micro_http as they're pretty much chaotic now.

`InvalidHeader` should be returned when the header is invalid and `Unsupported` should be returned when we found a header which we don't recognize and don't care about. The caller of `parse_header_line` should check the returned error and in case it is `Unsupported` the caller can ignore it, and when it is `InvalidHeader` it should return `BadRequest` or something similar and stop the parsing of the request.